Fix maude_decom.get_raw_aca_packets

javierggt opened the pull request:
## Description

This PR fixes what I would argue is a bug in `maude_decom.get_raw_aca_packets`. The problem was that I was calling this function with unusual arguments (frames inconsistent with the time range).

This function takes start/stop times and (optionally) the output from `maude.get_frames`. Using those frames, it should return all the full *ACA packets* within that time range (remember that the ACA telemetry comes split in four VCDU frames).

In the master version, it groups the packages using the VCDU frame counter, but instead of getting the counter from the frames themselves, it uses the requested time range to query MAUDE. Then it does not check whether the given frames match the frame counters from the time range, and proceeds happily, producing garbage (combining frames in groups of four, but starting at the wrong one).

There is no reason why this function should fail with inconsistent arguments, because the frame counter is in the frames themselves. By fixing this, I am also removing this extra call to MAUDE that is not necessary.

## Other changes

This PR includes two other changes I made while working on this. One was cause for another error. The other one was just a relatively simple change:
- Deal with the fact that sometimes IMGTYP can be 3 (code DNLD,  "memory download") and we do not want things to crash in this case. I made minimal changes so the output is filled and client code can decide what to do. I added a small unit test that just checks that the function does not crash.
- **Enable dynamic background MSIDs PIXTLM and BGDTYP for blob decom**, and change the type of PIXTLM and BGDTYP from int to string. This required changing one of the test data files.

## Interface impacts
None

## Testing
A new unit test was added for this (this PR is two commits, the first one is the test, so you can verufy the test fails before the change).

### Unit tests
<!-- At least one of these must be checked if unit tests exist. DELETE the unchecked/untested options. -->
- [x] Mac

Independent check of unit tests by Jean
- [x] Linux

### Functional tests
<!-- Describe and document results of any functional tests, otherwise leave the text below -->
I have been using this with aca_view (PR sot/aca_view/pull/169), and there I am making many functional tests.

jeanconn commented on chandra_aca/maude_decom.py:
```diff
@@ -310,7 +312,7 @@ def _aca_image_msid_list(pea):
     np.array([PIXEL_MAP_INV["6x6"][f"{k}1"] for k in _a2p]).T,
     np.array([PIXEL_MAP_INV["6x6"][f"{k}2"] for k in _a2p]).T,
-    [],
+    np.array([PIXEL_MAP_INV["4x4"][f"{k}1"] for k in _a2p]).T,
```
What does this change end up doing?

taldcroft replied:
Ping @javierggt for this question from @jeanconn .

jeanconn replied:
My cursory reading was that this was part of "Deal with the fact that sometimes IMGTYP can be 3 (code DNLD, "memory download")" by making more 4x4 images for that case instead of an empty list, but things were a little opaque in this part of the code so I figured I'd check.

javierggt replied:
ah, I missed the original question.

if image type is DNLD, the integer value is 3, which pulls that entry, which is empty in the master version. That ends up causing and exception. Wit this change DNLD images are treated just like 4x4 images.

jeanconn replied:
Thanks!

jeanconn commented on chandra_aca/maude_decom.py:
```diff
@@ -567,18 +573,27 @@ def get_raw_aca_packets(start, stop, maude_result=None, **maude_kwargs):
-        start=date_start,
-        stop=date_stop + stop_pad,
-        **maude_kwargs,
+    w = np.array([1, 2**8, 2**16])[::-1]
```
This part also probably deserves a comment?  The code is a little opaque here too.

jeanconn reviewed (approved):
Looks good to me and passed unit tests for me as well.  I left that optional suggestion to add a comment in a few of the harder-to-read pieces.

jeanconn commented on chandra_aca/maude_decom.py:
```diff
@@ -583,7 +589,10 @@ def get_raw_aca_packets(start, stop, maude_result=None, **maude_kwargs):
     frames = frames["frames"]
 
-    # also getting major and minor frames to figure out which is the first ACA packet in a group
+    # Getting major and minor frames to figure out which is the first ACA packet in a group
+    # The VCDU counter is stored in 24 bits, and struct.unpack does not have a 24-bit format code.
+    # We therefore decom the bytes separately and combine them into a single integer using the
+    # following weights:
```
Thanks!

javierggt commented:
I added some comments to address Jean's review comments.

While I was at it, I also changed it so the image values are masked if the image type is DNLD, and updated the test to check for that.

jeanconn reviewed (approved):
These updated changes seem fine to me and passed unit tests again for me.  I'd like to test out from aca_view but we can merge this now and do more aca_view testing later.
